{"url":"https://api.github.com/repos/microsoft/graspologic/issues/637","repository_url":"https://api.github.com/repos/microsoft/graspologic","labels_url":"https://api.github.com/repos/microsoft/graspologic/issues/637/labels{/name}","comments_url":"https://api.github.com/repos/microsoft/graspologic/issues/637/comments","events_url":"https://api.github.com/repos/microsoft/graspologic/issues/637/events","html_url":"https://github.com/microsoft/graspologic/issues/637","id":794492304,"node_id":"MDU6SXNzdWU3OTQ0OTIzMDQ=","number":637,"title":"[BUG] Graspologic Layouts can fail __init__.py if `umap` is installed","user":{"login":"daxpryce","id":899411,"node_id":"MDQ6VXNlcjg5OTQxMQ==","avatar_url":"https://avatars.githubusercontent.com/u/899411?v=4","gravatar_id":"","url":"https://api.github.com/users/daxpryce","html_url":"https://github.com/daxpryce","followers_url":"https://api.github.com/users/daxpryce/followers","following_url":"https://api.github.com/users/daxpryce/following{/other_user}","gists_url":"https://api.github.com/users/daxpryce/gists{/gist_id}","starred_url":"https://api.github.com/users/daxpryce/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/daxpryce/subscriptions","organizations_url":"https://api.github.com/users/daxpryce/orgs","repos_url":"https://api.github.com/users/daxpryce/repos","events_url":"https://api.github.com/users/daxpryce/events{/privacy}","received_events_url":"https://api.github.com/users/daxpryce/received_events","type":"User","site_admin":false},"labels":[{"id":1048646805,"node_id":"MDU6TGFiZWwxMDQ4NjQ2ODA1","url":"https://api.github.com/repos/microsoft/graspologic/labels/bug","name":"bug","color":"cded2d","default":true,"description":"Something isn't working"}],"state":"closed","locked":false,"assignee":null,"assignees":[],"milestone":null,"comments":3,"created_at":"2021-01-26T19:15:40Z","updated_at":"2021-02-09T17:36:31Z","closed_at":"2021-02-09T17:36:31Z","author_association":"CONTRIBUTOR","active_lock_reason":null,"body":"Apparently there are multiple libraries that use `umap` as their namespace. If the user has `umap` installed instead of our required `umap-learn`, environments can have unexpected results.\r\n\r\nThe actual fix for this is to `pip uninstall umap umap-learn && pip install umap-learn` to purge umap from your environment.\r\n\r\nAn argument can be made that since this is an easy trap to fall into, we should do something to alert our users that their environment isn't what we (or possibly they) expected.\r\n\r\nAnother argument could be that we should create an FAQ wiki page about problems just like this so we can tell people how to fix them going forward.\r\n\r\nI'm not convinced we want a coding solution, but I wanted to spur a discussion about this.\r\n\r\n## Expected Behavior\r\n`import graspologic` should succeed\r\n\r\n## Actual Behavior\r\n`import graspologic` fails when `umap` is installed instead of `umap-learn`.\r\n\r\n## Full Traceback\r\n```pytb\r\nTraceback (most recent call last):\r\n  File \"/app/d3m/d3m/index.py\", line 174, in load_all\r\n    get_primitive(primitive_path)\r\n  File \"/app/d3m/d3m/index.py\", line 117, in get_primitive\r\n    return getattr(module, name)\r\n  File \"/app/d3m/d3m/namespace.py\", line 116, in __getattr__\r\n    primitive = entry_point.resolve()\r\n  File \"/usr/local/lib/python3.6/dist-packages/pkg_resources/__init__.py\", line 2459, in resolve\r\n    module = __import__(self.module_name, fromlist=['__name__'], level=0)\r\n  File \"/app/primitives-interfaces/jhu_primitives/__init__.py\", line 26, in <module>\r\n    from .sgm_nomination import SgmNomination\r\n  File \"/app/primitives-interfaces/jhu_primitives/sgm_nomination/__init__.py\", line 2, in <module>\r\n    from .sgm_nomination import SgmNomination\r\n  File \"/app/primitives-interfaces/jhu_primitives/sgm_nomination/sgm_nomination.py\", line 12, in <module>\r\n    from graspologic.match import GraphMatch\r\n  File \"/usr/local/lib/python3.6/dist-packages/graspologic/__init__.py\", line 9, in <module>\r\n    import graspologic.layouts\r\n  File \"/usr/local/lib/python3.6/dist-packages/graspologic/layouts/__init__.py\", line 7, in <module>\r\n    from .auto import layout_tsne, layout_umap\r\n  File \"/usr/local/lib/python3.6/dist-packages/graspologic/layouts/auto.py\", line 11, in <module>\r\n    import umap\r\n  File \"/usr/local/lib/python3.6/dist-packages/umap/__init__.py\", line 2, in <module>\r\n    from .umap_ import UMAP\r\n  File \"/usr/local/lib/python3.6/dist-packages/umap/umap_.py\", line 30, in <module>\r\n    import umap.distances as dist\r\nAttributeError: module 'umap' has no attribute 'distances'\r\n\r\n```\r\n@alyakin314 was the one who found this and I'm sure he has some thoughts on which direction we should go with this.\r\n","closed_by":{"login":"daxpryce","id":899411,"node_id":"MDQ6VXNlcjg5OTQxMQ==","avatar_url":"https://avatars.githubusercontent.com/u/899411?v=4","gravatar_id":"","url":"https://api.github.com/users/daxpryce","html_url":"https://github.com/daxpryce","followers_url":"https://api.github.com/users/daxpryce/followers","following_url":"https://api.github.com/users/daxpryce/following{/other_user}","gists_url":"https://api.github.com/users/daxpryce/gists{/gist_id}","starred_url":"https://api.github.com/users/daxpryce/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/daxpryce/subscriptions","organizations_url":"https://api.github.com/users/daxpryce/orgs","repos_url":"https://api.github.com/users/daxpryce/repos","events_url":"https://api.github.com/users/daxpryce/events{/privacy}","received_events_url":"https://api.github.com/users/daxpryce/received_events","type":"User","site_admin":false},"reactions":{"url":"https://api.github.com/repos/microsoft/graspologic/issues/637/reactions","total_count":0,"+1":0,"-1":0,"laugh":0,"hooray":0,"confused":0,"heart":0,"rocket":0,"eyes":0},"timeline_url":"https://api.github.com/repos/microsoft/graspologic/issues/637/timeline","performed_via_github_app":null,"state_reason":"completed"}